src/main/java/com/github/manolo8/darkbot/core/objects/swf/FlashDictionary.java

```diff
@@ -0,0 +1,161 @@
+        return this;
+    }
+
+    public FlashDictionary<K, V> setIgnoreEmpty(boolean ignoreEmpty) {
+        this.ignoreEmpty = ignoreEmpty;
+        return this;
+    }
+
+    public void addListener(K key, Consumer<V> consumer) {
+        addListener(key, consumer, true);
+    }
+
+    public void addListener(K key, Consumer<V> consumer, boolean cacheValue) {
+        this.listeners.computeIfAbsent(key, k -> (cacheValue ? new Lazy<>() : new Lazy.NoCache<>())).add(consumer);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void update() {
+        if (address == 0 || (ignoreEmpty && listeners.isEmpty())) return;
+
+        long hashTable = API.readLong(address + 32); //todo get table offset from script object traits
+        if (hashTable == 0) return;
+
+        long atoms = API.readLong(hashTable + 8);
+        long sizeAndExp = API.readLong(hashTable + 16);
+
+        //boolean hasIterIndex = (atoms & 0x04) != 0; //unused here
+
+        int size = (int) sizeAndExp;
+        int exp = (int) (sizeAndExp >> 32);
+
+        int capacity = (1 << (exp - 1)) * Long.BYTES;
+
+        if (size <= 0 || size > 1024 || capacity <= 0 || capacity > 1024 * 2 * 8) return;
+        if (entries.length < size) entries = Arrays.copyOf(entries, size);
+
+        this.size = size;
+
+        atoms = (atoms & ByteUtils.ATOM_MASK) + 8; //remove tags & skip c++ vtable
+
+        for (int offset = 0, idx = 0; offset < capacity && idx < size; offset += DataReader.MAX_CHUNK_SIZE) {
+            try (DataReader r = API.readData(atoms + offset,
+                    Math.min(DataReader.MAX_CHUNK_SIZE, capacity - offset))) {
+
+                while (r.getAvailable() >= 16 && idx < size) {
+                    long key = r.getLong();
+                    if (AtomKind.isNullAtom(key)) continue; // key cannot be null
+
+                    AtomKind keyKind = AtomKind.of(key);
+                    if (keyKind != this.keyKind) continue;
+
+                    Entry<K, V> entry = entries[idx];
+                    if (entry == null) entries[idx] = entry = new Entry<>(listeners);
+
+                    entry.set((K) keyKind.read(key), (V) valueKind.read(r.getLong()));
+                    idx++;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void update(long address) {
+        super.update(address);
+        if (autoUpdatable) update();
+    }
+
+    @Override
+    public int getSize() {
+        return Math.min(size, entries.length);
+    }
+
+    @Override
+    public long getPtr(int i) {
+        if (i < getSize()) {
+            Entry entry = entries[i];
+            if (entry != null && entry.value instanceof Long)
+                return (long) entry.value;
+        }
+
+        return 0;
+    }
+
+    private static class Entry<K, V> {
+        private final Map<K, Lazy<V>> listeners;
+
+        private K key;
+        private V value;
+
+        public Entry(Map<K, Lazy<V>> listeners) {
```

Comment:
If you need to pass down the listeners to the entry, it may be simpler to just make this class non-static and use them from the outer class directly